lib/puppet/provider/snmp_notification_receiver/nxapi.rb

```diff
@@ -0,0 +1,164 @@
+      version = 'v1'
+    elsif v.version.eql?('2c')
+      version = 'v2'
+    elsif v.version.eql?('3')
+      version = 'v3'
+    else
+      version = nil
+    end
+
+    current_state = {
+      ensure:           :present,
+      name:             snmpnotificationreceiver_name,
+      port:             v.port,
+      username:         v.username,
+      version:          version,
+      type:             v.type,
+      security:         v.security,
+      vrf:              v.vrf,
+      source_interface: v.source_interface,
+    }
+    new(current_state)
+  end # self.properties_get
+
+  def self.instances
+    snmpnotificationreceivers = []
+    Cisco::SnmpNotificationReceiver.receivers.each do |snmpnotificationreceiver_name, v|
+      snmpnotificationreceivers << properties_get(snmpnotificationreceiver_name, v)
+    end
+
+    snmpnotificationreceivers
+  end
+
+  def self.prefetch(resources)
+    snmpnotificationreceivers = instances
+
+    resources.keys.each do |id|
+      provider = snmpnotificationreceivers.find { |snmpnotificationreceiver| snmpnotificationreceiver.name.to_s == id.to_s }
+      resources[id].provider = provider unless provider.nil?
+    end
+  end # self.prefetch
+
+  def exists?
+    @property_hash[:ensure] == :present
+  end
+
+  def create
+    @property_flush[:ensure] = :present
+  end
+
+  def destroy
+    @property_flush[:ensure] = :absent
+  end
+
+  def validate
+    required = []
+    REQUIRED_PROPS.each do |prop|
+      required << prop unless @resource[prop]
+    end
+    fail ArgumentError,
+         "You must specify the following properties: #{required}" \
+           unless required.empty?
+
+    fail ArgumentError,
+         "The 'type' property only supports a setting of 'traps' when 'version' is set to '1'" \
+            if !@resource[:type].eql?('traps') && @resource[:version].eql?(:v1)
+
+    fail ArgumentError,
+         "The 'security' property is only supported when 'version' is set to '3'" \
+            if @resource[:security] && !@resource[:version].eql?(:v3)
+  end
+
+  def munge_flush(val)
+    if val.is_a?(String) && val.eql?('unset')
+      nil
+    elsif val.is_a?(Integer) && val.eql?(-1)
+      nil
+    elsif val.is_a?(Symbol) && val.eql?(:true)
+      true
+    elsif val.is_a?(Symbol) && val.eql?(:false)
+      false
+    elsif val.is_a?(Symbol)
+      val.to_s
+    elsif val.is_a?(Integer)
+      val.to_s
+    else
+      val
+    end
+  end
+
+  def munge_version(_va)
+    if @resource[:version].eql?(:v1)
+      '1'
+    elsif @resource[:version].eql?(:v2)
+      '2c'
+    elsif @resource[:version].eql?(:v3)
+      '3'
+    end
+  end
+
+  def flush # rubocop:disable Metrics/CyclomaticComplexity
+    @snmpnotificationreceiver.destroy if @snmpnotificationreceiver
+    @snmpnotificationreceiver = nil
+
+    return if @property_flush[:ensure] == :absent
+
+    validate
+
+    # Needed due to NXOS issue regarding removing and adding resources too quickly
+    sleep 2
```

Comment:
@jonnytpuppet I did not realize that it was accepted for snmp_user but the issue still remains.  We really should fix this in both places.  In my experience on cisco devices, any time we rely on sleep statements for automation bad things happen where timing can differ across various cisco devices.  We wind up having to change and tweak it later.  Better to write once and build in the flexibility to handle platform differences.

I would loop a predetermined number of times (5-7 for example), sleep 1 second between checks and break out of the loop the second the desired state is reached.  Generate an error if we complete all loops and don't reach the desired state.